packages/grafana-data/src/types/logs.ts

```diff
@@ -260,3 +260,37 @@ export const hasLogsContextUiSupport = (datasource: unknown): datasource is Data
 
   return withLogsSupport.getLogRowContextUi !== undefined;
 };
+
+export interface AnalyzeQueryOptions {
+  check: string;
+  attributes: KeyValue<string>;
+}
+
+/**
+ * Data sources that support query manipulation through `modifyQuery` and `analyzeQuery` in Explore.
+ * Allows for interactions such as changing the query from logs details, or displaying filter status.
+ * @internal
+ */
+export interface DataSourceWithQueryManipulationSupport<TQuery extends DataQuery, TAnalyzeQueryResult = boolean> {
+  /**
+   * Used in explore
+   */
+  modifyQuery(query: TQuery, action: QueryFixAction): TQuery;
+
+  /**
+   * Used in explore for Log details
+   *
+   * @alpha
+   */
+  analyzeQuery?(query: TQuery, options: AnalyzeQueryOptions): TAnalyzeQueryResult;
+}
+
+/**
+ * @internal
+ */
+export const hasQueryManipulationSupport = <TQuery extends DataQuery>(
+  datasource: unknown,
+  method: keyof DataSourceWithQueryManipulationSupport<TQuery>
+): datasource is DataSourceWithQueryManipulationSupport<TQuery> => {
+  return datasource !== null && typeof datasource === 'object' && method in datasource;
```

Comment:
I think that here we should check for all required methods instead of passing method name. Cause if I would use `const hasSupport = hasQueryManipulationSupport(datasource, 'analyzeQuery')` and my data source has `analyzeQuery` method, but it does not have  `modifyQuery`, it will produce incorrect result. 

So maybe just something like
```
export const hasQueryManipulationSupport = <TQuery extends DataQuery>(
  datasource: unknown,
): datasource is DataSourceWithQueryManipulationSupport<TQuery> => {
  return datasource !== null && typeof datasource === 'object' && datasource.modifyQuery !- undefined
```

Reply:
Good point! Changed. Also took the opporunity to make `analyzeQuery` not optional, as `modifyQuery` is.

Reply:
Split into two interfaces 👍